lib/Reader/readerir.cpp

```diff
@@ -4457,6 +4457,42 @@ void GenIR::dup(IRNode *Opr, IRNode **Result1, IRNode **Result2) {
   *Result2 = Opr;
 }
 
+bool GenIR::interlockedCmpXchg(IRNode *Destination, IRNode *Exchange,
+                               IRNode *Comparand, IRNode **Result,
+                               CorInfoIntrinsics IntrinsicID) {
+  ASSERT(Exchange->getType() == Comparand->getType());
+  switch (IntrinsicID) {
+  case CORINFO_INTRINSIC_InterlockedCmpXchg32:
+    ASSERT(Exchange->getType() == Type::getInt32Ty(*JitContext->LLVMContext));
+    break;
+  case CORINFO_INTRINSIC_InterlockedCmpXchg64:
+    ASSERT(Exchange->getType() == Type::getInt64Ty(*JitContext->LLVMContext));
+    break;
+  default:
+    throw NotYetImplementedException("interlockedCmpXchg");
+  }
+
+  Type *ComparandTy = Comparand->getType();
+  Type *ComparandPtrTy = getUnmanagedPointerType(ComparandTy);
+  Type *DestinationTy = Destination->getType();
+
+  if (DestinationTy->isIntegerTy()) {
+    Destination =
+        (IRNode *)LLVMBuilder->CreateIntToPtr(Destination, ComparandPtrTy);
+  } else {
+    ASSERT(DestinationTy->isPointerTy());
+    Destination =
+        (IRNode *)LLVMBuilder->CreatePointerCast(Destination, ComparandPtrTy);
```

Comment:
I don't think this is correct--if the destination is a managed pointer (i.e. in `addrspace(1)`), this will end up casting to `addrspace(0)`. All we want to do is cast the destination's element type while leaving the address space alone.

Also, in the case above that where destination type is an integral type, I think that we have to be conservative and treat it as a managed pointer.